mk2/launchpad.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2013 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package launchpad provides interfaces to talk to
+// Novation Launchpads via MIDI in and out.
+package mk2
+
+import (
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/rakyll/portmidi"
+)
+
+// Launchpad represents a device with an input and output MIDI stream.
+type Launchpad struct {
+	inputStream  *portmidi.Stream
+	outputStream *portmidi.Stream
+}
+
+// Hit represents physical touches to Launchpad buttons.
+type Hit struct {
+	X int
+	Y int
+}
+
+// Open opens a connection Launchpad and initializes an input and output
+// stream to the currently connected device. If there are no
+// devices are connected, it returns an error.
+func Open() (*Launchpad, error) {
+	input, output, err := discover()
+	if err != nil {
+		return nil, err
+	}
+
+	var inStream, outStream *portmidi.Stream
+	if inStream, err = portmidi.NewInputStream(input, 1024); err != nil {
+		return nil, err
+	}
+	if outStream, err = portmidi.NewOutputStream(output, 1024, 0); err != nil {
+		return nil, err
+	}
+	return &Launchpad{inputStream: inStream, outputStream: outStream}, nil
+}
+
+// Listen listens the input stream for hits.
+func (l *Launchpad) Listen() <-chan Hit {
+	ch := make(chan Hit)
+	go func(pad *Launchpad, ch chan Hit) {
+		for {
+			// sleep for a while before the new polling tick,
+			// otherwise operation is too intensive and blocking
+			time.Sleep(10 * time.Millisecond)
+			hits, err := pad.Read()
+			if err != nil {
+				continue
+			}
+			for i := range hits {
+				ch <- hits[i]
+			}
+		}
+	}(l, ch)
+	return ch
+}
+
+// Read reads hits from the input stream. It returns max 64 hits for each read.
+func (l *Launchpad) Read() (hits []Hit, err error) {
+	var evts []portmidi.Event
+	if evts, err = l.inputStream.Read(1024); err != nil {
+		return
+	}
+	for _, evt := range evts {
+		if evt.Data2 > 0 {
+			var x, y int64
+			if evt.Status == 176 {
+				// top row button
+				// FIXME
+				x = evt.Data1 - 104
+				y = -1
+			} else {
+				x = evt.Data1 % 10
+				y = (evt.Data1 - x) / 10
+			}
+			hits = append(hits, Hit{X: int(x), Y: int(y)})
+		}
+	}
+	return
+}
+
+// Light lights the button at x,y with the given red, green, and blue values.
+// x and y are [0, 7], r, g, and b are [0, 63]
+func (l *Launchpad) Light(x, y, r, g, b int) error {
```

Comment:
Mkay. I wanted to avoid making my own types but this is probably easier to understand. Probably could be reused for a future GetColor() method etc.